2.0/m_stun.cpp

```diff
@@ -0,0 +1,332 @@
+			}
+		}
+		else
+		{
+			// Adding - XXX todo make this respect <insane> tag perhaps..
+			long duration;
+			std::string expr;
+			if (parameters.size() > 2)
+			{
+				duration = ServerInstance->Duration(parameters[1]);
+				expr = parameters[2];
+			}
+			else
+			{
+				duration = 0;
+				expr = parameters[1];
+			}
+
+			Stun* r = new Stun(ServerInstance->Time(), duration, user->nick.c_str(), expr.c_str(), target.c_str());
+			if (ServerInstance->XLines->AddLine(r, user))
+			{
+				if (!duration)
+				{
+					ServerInstance->SNO->WriteToSnoMask('x',"%s added permanent STUN for %s: %s",
+						user->nick.c_str(), target.c_str(), expr.c_str());
+				}
+				else
+				{
+					time_t c_requires_crap = duration + ServerInstance->Time();
+					std::string timestr = ServerInstance->TimeString(c_requires_crap);
+					ServerInstance->SNO->WriteToSnoMask('x', "%s added timed STUN for %s to expire on %s: %s",
+						user->nick.c_str(), target.c_str(), timestr.c_str(), expr.c_str());
+				}
+			}
+			else
+			{
+				delete r;
+				user->WriteServ("NOTICE %s :*** Stun for %s already exists", user->nick.c_str(), target.c_str());
+				return CMD_FAILURE;
+			}
+		}
+		return CMD_SUCCESS;
+	}
+
+	RouteDescriptor GetRouting(User* user, const std::vector<std::string>& parameters)
+	{
+		if (IS_LOCAL(user))
+			return ROUTE_LOCALONLY; // spanningtree will send ADDLINE
+
+		return ROUTE_BROADCAST;
+	}
+};
+
+class ModuleStun : public Module
+{
+	CommandStun cmd;
+	StunFactory f;
+	bool affectopers;
+
+	std::string deaf_bypasschars;
+	std::string deaf_bypasschars_uline;
+
+ public:
+	ModuleStun() : cmd(this)
+	{
+	}
+
+	void init()
+	{
+		ServerInstance->XLines->RegisterFactory(&f);
+		ServerInstance->Modules->AddService(cmd);
+
+		Implementation eventlist[] = { I_OnStats, I_OnUserPreMessage, I_OnUserPreNotice, I_OnRehash };
+		ServerInstance->Modules->Attach(eventlist, this, sizeof(eventlist)/sizeof(Implementation));
+		OnRehash(NULL);
+	}
+
+	virtual ~ModuleStun()
+	{
+		ServerInstance->XLines->DelAll("STUN");
+		ServerInstance->XLines->UnregisterFactory(&f);
+	}
+
+	/*
+	 * I dont think i need this for OnUserPreNotice and OnUserPreMessage...?
+	 *
+	void Prioritize()
+	{
+		Module* alias = ServerInstance->Modules->Find("m_alias.so");
+		ServerInstance->Modules->SetPriority(this, I_OnPreCommand, PRIORITY_BEFORE, &alias);
+	}
+	*/
+
+	virtual ModResult OnStats(char symbol, User* user, string_list& out)
+	{
+		if (symbol != 'K')
+			return MOD_RES_PASSTHRU;
+
+		ServerInstance->XLines->InvokeStats("STUN", 223, user, out);
+		return MOD_RES_DENY;
+	}
+
+	virtual void OnRehash(User* user)
+	{
+		ConfigTag* tag = ServerInstance->Config->ConfValue("stun");
+
+		deaf_bypasschars = tag->getString("bypasschars");
+		deaf_bypasschars_uline = tag->getString("bypasscharsuline");
+
+		affectopers = tag->getBool("affectopers", false);
+	}
+
+	/*
+	 * The rest of the code is mostly taken from m_deaf.so
+	 * Copyright (C) 2006-2007 Dennis Friis <peavey@inspircd.org>
```

Comment:
If the copyright  is already mentioned in the header you don't need to repeat it.